Source/WebGPU/WGSL/Parser.cpp

```diff
@@ -329,6 +332,45 @@ Expected<UniqueRef<AST::TypeDecl>, Error> Parser<Lexer>::parseTypeDeclAfterIdent
     RETURN_NODE_REF(NamedType, WTFMove(name));
 }
 
+// https://www.w3.org/TR/WGSL/#syntax-array_type_decl
+// ArrayTypeDecl:
+//      'array' ('<' TypeDecl (',' ElementCountExpression)? '>')?
+// ElementCountExpression:
+//      AdditiveExpression | BitwiseExpression
+template<typename Lexer>
+Expected<UniqueRef<AST::TypeDecl>, Error> Parser<Lexer>::parseArrayType()
+{
+    START_PARSE();
+
+    CONSUME_TYPE(KeywordArray);
+
+    std::unique_ptr<AST::TypeDecl> maybeElementType;
+    std::unique_ptr<AST::Expression> maybeElementCount;
+
+    if (current().m_type == TokenType::LT) {
+        // We differ from the WGSL grammar here by allowing the type to be optional,
+        // which allows us to use `parseArrayType` in `parseCallableExpression`.
+        consume();
+
+        PARSE(elementType, TypeDecl);
+        maybeElementType = elementType.moveToUniquePtr();
+
+        if (current().m_type == TokenType::Comma) {
+            consume();
+            // FIXME: According to https://www.w3.org/TR/WGSL/#syntax-element_count_expression
+            // this should be: AdditiveExpression | BitwiseExpression.
+            //
+            // The WGSL grammar doesn't specify expression operator precedence so
```

Comment:
Link to spec issue?

Reply:
This is more of an issue with our lack of support for parsing expressions.